### PR TITLE
security: add certificate revocation record

### PR DESCRIPTION
**Pull request by Harsh23Kashyap**
## Summary
- Add low-level `Revoke` and `IsRevoked` certificate revocation APIs in `std/security`.
- Store revocation entries as process-local, mutex-protected records derived from `UCLA-IRL/ndnrevoke`'s revocation record shape.
- Derive `REVOKE`-style record names from certificate names and keep record metadata including the certificate name, public key hash, reason, and timestamp.
- Add tests for nil handling, unrelated certificates, repeated revocation checks, and concurrent access.

## Test plan
- `go test ./std/security -run 'TestCertificateRevocationRecord' -count=20`
- `go test -race ./std/security`
- `go test ./std/security`
- `make test`

### CODE REVIEW

**Comment by tianyuan129** [comment]
https://github.com/UCLA-IRL/ndnrevoke.

Please use this as reference for Revocation Record.

**Comment by Harsh23Kashyap** [comment]
Thanks for the pointer. I looked through `UCLA-IRL/ndnrevoke`, especially the record/revoker/checker flow, and updated this PR to make the local revocation record closer to that model.

The public API is still limited to the requested low-level surface:

```go
Revoke(cert ndn.Data)
IsRevoked(cert ndn.Data) bool
```

Internally, `Revoke` now builds and stores a revocation record instead of just a bare certificate-name set entry. The record derives a `REVOKE`-style name from the certificate name, stores the original cert name, public key hash, reason, and timestamp, and `IsRevoked` checks using that derived record name. I kept ledger/checker integration out of this PR since this still looks like the low-level API step before the app-facing flow.

I reran:

```bash
go test ./std/security -run 'TestCertificateRevocationRecord' -count=20
go test -race ./std/security
go test ./std/security
make test
```

All passed.

**Comment by tianyuan129** [comment]
Thanks for the initial work! I saw that you currently have a struct storing revocation records and make the code checks against that. That's exactly the next step we want to work on.

The library has a module [trust_config](https://github.com/named-data/ndnd/blob/main/std/security/trust_config.go) that automates the data authentication and authorization workflow, and actually the downstream applications who uses ndnd mainly rely on the high-level APIs that are binded to this module to get rid of the crypto details.

At high level, it executes the NDN trust schema and recursively climb the certification chain for a received data until reaching a bootstrapped (or otherwise externally authenticated) trust anchor. We need corporate the revocation here -- so obviously we need to store received revocation records here....but first lets think about design space.

**Comment by Harsh23Kashyap** [comment]
Thanks for the context. I’ll pause on further implementation for now and read through `trust_config` and the papers first, so I understand where revocation should fit in the existing validation flow before making the next change.

**Comment by Harsh23Kashyap** [comment]
I went through the `trust_config` validation path and the two papers at a high level.

My understanding is that `trust_config` is the right place to think about the next step, because it already handles the high-level data validation flow: checking the trust schema, fetching certificates, using the certificate cache, and recursively climbing the certificate chain until a trust anchor.

I also see why the Envelope-related code can look like it jumps between trust anchors. I’ll avoid touching that part for now unless needed, and focus only on where revocation records should be stored and checked inside the current validation workflow.

**Comment by Harsh23Kashyap** [comment]
Based on my current understanding, I see the next step as a design problem around `TrustConfig`, not just the low-level `Revoke` / `IsRevoked` helpers.

The current PR gives us the low-level revocation record primitive. The next question is where that record should live and when `trust_config` should check it during validation.

I think the design space is:

1. Local `TrustConfig` revocation store first: store received revocation records inside `TrustConfig`, and reject a certificate before it is used for validation. This would cover certificates coming from cache, fetch, or direct validation.

2. Store revocation record Data, not just names: instead of only keeping a cert name or derived record name, store the received revocation record Data/metadata so we can preserve reason, timestamp, public key hash, revoker, and later signature validation state.

3. Network or ledger check later: during validation, `TrustConfig` can derive the `REVOKE` record name and fetch either a revocation record or a nack, closer to `ndnrevoke`. I think this is larger and should probably come after the local store design is clear.

So my proposed next step is to start with (1), while shaping the store so (2) can fit naturally later. I’ll keep the Envelope-related code untouched for now and only use the Envelope paper as context for why revocation belongs as a validation/checker step.

Does this direction sound right before I update the PR further?

**Comment by Harsh23Kashyap** [comment]
Sharing the HLD and diagram for the TrustConfig revocation design discussed above.

**Attachments**
- [Design gist (draw.io + HLD markdown)](https://gist.github.com/Harsh23Kashyap/db05710ceb0298acb27658db134d85a1)
- [trustconfig-revocation-flow.drawio](https://gist.githubusercontent.com/Harsh23Kashyap/db05710ceb0298acb27658db134d85a1/raw/trustconfig-revocation-flow.drawio)
- [trustconfig-revocation-hld.md](https://gist.githubusercontent.com/Harsh23Kashyap/db05710ceb0298acb27658db134d85a1/raw/trustconfig-revocation-hld.md)

## HLD

**Goal:** integrate revocation into the high-level validation workflow, not only as the low-level `Revoke` / `IsRevoked` helpers in this PR.

**Proposed flow**

```mermaid
flowchart TD
  dataPacket["Received Data"] --> validate["TrustConfig.Validate"]
  validate --> keyLocator["Read KeyLocator"]
  keyLocator --> certSource["Cert from cache, fetch, or direct args"]
  certSource --> revokedCheck["Check TrustConfig revocation store"]
  revokedCheck -->|"revoked"| reject["Fail validation"]
  revokedCheck -->|"not revoked"| schemaCheck["Trust schema check"]
  schemaCheck --> sigCheck["Signature check"]
  sigCheck --> certChain["Validate signing cert"]
  certChain --> validate
```

**Where to check revocation**
- before using a cached certificate
- after fetching a certificate, before caching it
- before using `args.cert` for signature validation
- when validating a certificate packet itself (`ContentTypeKey`)

**Sequenced design space**
1. Local `TrustConfig` revocation store first
2. Store revocation record Data/metadata next
3. Network or ledger lookup later, closer to `ndnrevoke`

**Out of scope for now**
- network revocation lookup
- Envelope-like code changes
- ledger/checker integration

The draw.io diagram shows the same flow with cert sources, fail path, trust-anchor path, and notes on where the store should live.

**Comment by Harsh23Kashyap** [comment]
Following up on the design proposal above — no rush, just checking whether the direction makes sense before I start coding.

To summarize what I'd do first if this looks reasonable:
- add a revocation store scoped to `TrustConfig`
- check it inside `Validate` before a cert is used from cache, fetch, or direct args
- keep the current `Revoke` / `IsRevoked` helpers as the low-level primitive for now

I'll hold off on further PR changes until I hear back. Happy to adjust the plan if you'd prefer a different starting point (e.g. storing received revocation record Data earlier, or a different hook in the validation path).

**Comment by tianyuan129** [comment]
> Based on my current understanding, I see the next step as a design problem around `TrustConfig`, not just the low-level `Revoke` / `IsRevoked` helpers.
> 
> The current PR gives us the low-level revocation record primitive. The next question is where that record should live and when `trust_config` should check it during validation.
> 
> I think the design space is:
> 
> 1. Local `TrustConfig` revocation store first: store received revocation records inside `TrustConfig`, and reject a certificate before it is used for validation. This would cover certificates coming from cache, fetch, or direct validation.
> 2. Store revocation record Data, not just names: instead of only keeping a cert name or derived record name, store the received revocation record Data/metadata so we can preserve reason, timestamp, public key hash, revoker, and later signature validation state.
> 3. Network or ledger check later: during validation, `TrustConfig` can derive the `REVOKE` record name and fetch either a revocation record or a nack, closer to `ndnrevoke`. I think this is larger and should probably come after the local store design is clear.
> 
> So my proposed next step is to start with (1), while shaping the store so (2) can fit naturally later. I’ll keep the Envelope-related code untouched for now and only use the Envelope paper as context for why revocation belongs as a validation/checker step.
> 
> Does this direction sound right before I update the PR further?

For storing revocation records, I am thinking probably we don't need a separate store. The same store for certificate works for revocation too, since both are data, and store support name-based query. For a certificate, it's straightforward to use naming convention to construct its potential revocation name and query the store.

Yeah now let's assume revocation records are already delivered to local...we need an API to install revocation records.
Technically, the engine _can_ express an Interest to network to ask for revocation, but I feel we should start simple...but  we definitely need leave comments in code.

Revocation records has a notBefore -- marking data produced before this timestamp should be valid. It's related to #186, which I am trying to review and merge now. We need to implement the timestamp checking logic after we merged the feature (soon).

**Comment by tianyuan129** [comment]
> Sharing the HLD and diagram for the TrustConfig revocation design discussed above.
> 
> **Attachments**
> 
> * [Design gist (draw.io + HLD markdown)](https://gist.github.com/Harsh23Kashyap/db05710ceb0298acb27658db134d85a1)
> * [trustconfig-revocation-flow.drawio](https://gist.githubusercontent.com/Harsh23Kashyap/db05710ceb0298acb27658db134d85a1/raw/trustconfig-revocation-flow.drawio)
> * [trustconfig-revocation-hld.md](https://gist.githubusercontent.com/Harsh23Kashyap/db05710ceb0298acb27658db134d85a1/raw/trustconfig-revocation-hld.md)
> 
> ## HLD
> **Goal:** integrate revocation into the high-level validation workflow, not only as the low-level `Revoke` / `IsRevoked` helpers in this PR.
> 
> **Proposed flow**
> 
> ```mermaid
> flowchart TD
>   dataPacket["Received Data"] --> validate["TrustConfig.Validate"]
>   validate --> keyLocator["Read KeyLocator"]
>   keyLocator --> certSource["Cert from cache, fetch, or direct args"]
>   certSource --> revokedCheck["Check TrustConfig revocation store"]
>   revokedCheck -->|"revoked"| reject["Fail validation"]
>   revokedCheck -->|"not revoked"| schemaCheck["Trust schema check"]
>   schemaCheck --> sigCheck["Signature check"]
>   sigCheck --> certChain["Validate signing cert"]
>   certChain --> validate
> ```
> **Where to check revocation**
> 
> * before using a cached certificate
> * after fetching a certificate, before caching it
> * before using `args.cert` for signature validation
> * when validating a certificate packet itself (`ContentTypeKey`)
> 
> **Sequenced design space**
> 
> 1. Local `TrustConfig` revocation store first
> 2. Store revocation record Data/metadata next
> 3. Network or ledger lookup later, closer to `ndnrevoke`
> 
> **Out of scope for now**
> 
> * network revocation lookup
> * Envelope-like code changes
> * ledger/checker integration
> 
> The draw.io diagram shows the same flow with cert sources, fail path, trust-anchor path, and notes on where the store should live.

LGTM. Not sure if your diagram captures that, but we need to skip trust anchor for revocation checking.

**Comment by Harsh23Kashyap** [comment]
Thanks for the feedback — this helps a lot. A few notes on how I'm reading it and what I'll do next.

**On storage:** I'll drop the idea of a separate `TrustConfig` revocation store. Revocation records are Data like certificates, so reusing the existing name-based store (via `keychain.Store()`) makes sense — derive the `REVOKE` name from the cert and query the store. I'll update the HLD/diagram to reflect that.

**On the install API:** Agreed that the next concrete piece is an API to install received revocation records locally (assuming they're already delivered). I'll shape that before wiring checks into `Validate`.

**On validation hooks:** I'll keep the same hook points from the HLD (cache hit, post-fetch before cache insert, direct `args.cert`, cert packet validation), with one addition from your comment: **skip revocation checking for trust anchors**.

**On `notBefore`:** I'll hold off on timestamp-based revocation logic until #186 lands, then wire it in as a follow-up.

**On network fetch:** I'll start with local install + store lookup only, and leave comments in code where a future Interest-based revocation fetch could plug in.

**Plan for the next PR update:**
1. Add install API for revocation record Data into the shared store
2. During `Validate`, derive `REVOKE` name and query store before using a cert (non-anchor only)
3. Update the diagram to show shared store + trust-anchor skip
4. Tests for install + rejected validation when a revocation record is present

I'll start on this unless you'd rather split install vs. validation into separate PRs.

**Comment by tianyuan129** [comment]
> **On `notBefore`:** I'll hold off on timestamp-based revocation logic until #186 lands, then wire it in as a follow-up.

Given #186 has already been merged, you can use the timestamp from Data signature now.

**Comment by Harsh23Kashyap** [comment]
Pushed the TrustConfig integration in `0e2c8b2`:

- `InstallRevocationRecord` stores revocation record Data in `keychain.Store()`
- `Validate` checks the derived `REVOKE` name before using a cert (cache / fetch / direct / cert packet paths)
- Trust anchors are skipped for revocation checks
- `TODO(revocation)` left for future network Interest fetch; `notBefore` deferred until #186

Tests: `go test -race ./std/security`, `make test` — all green.

**Comment by Harsh23Kashyap** [comment]
Addressing the feedback that revocation records should be Data packets (not in-memory structs):

- `MakeRevocationRecord` builds a proper revocation record Data packet with ndnrevoke-style content TLVs (timestamp, reason, public key hash)
- `Revoke` now stores the encoded Data wire instead of a Go struct
- `InstallRevocationRecord` validates both REVOKE naming and record content
- TrustConfig store lookup path unchanged

Tests pass (`go test -race ./std/security`, `make test`).

File: std/security/certificate.go [hunk 1]
```diff
@@ -33,7 +59,6 @@ type SignCertArgs struct {
 // SignCert signs a new NDN certificate with the given signer.
 // Data must have either a Key or Secret in the Content.
 func SignCert(args SignCertArgs) (enc.Wire, error) {
-	// Check all parameters (strict for certs)
```
Comment (tianyuan129) [inline std/security/certificate.go:36]: why you delete those comments?

File: std/security/certificate.go [hunk 2]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
 	return false
 }
 
+// Revoke records a certificate as revoked by storing a revocation record Data packet.
+func Revoke(cert ndn.Data) {
```
Comment (tianyuan129) [inline std/security/certificate.go]: We should make the function naming consistent `SignCert `vs. `RevokeCert`

File: std/security/certificate.go [hunk 3]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
...
+	if err != nil {
+		return
+	}
+
+	data, _, err := spec.Spec{}.ReadData(enc.NewWireView(wire))
+	if err != nil {
+		return
+	}
+
+	_ = revocationStore.Put(data.Name(), wire.Join())
```
Comment (tianyuan129) [inline std/security/certificate.go]: This function is stateless. I believe all functions in this file should be stateless, keep all states in `TrustConfig`.

BTW `SignCert` is stateless too.

File: std/security/certificate.go [hunk 4]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
...
+	encoded, err := spec.Spec{}.MakeData(recordName, cfg, content, args.Signer)
+	if err != nil {
+		return nil, err
+	}
+
+	return encoded.Wire, nil
+}
+
+// RevocationRecordName derives the REVOKE-style record name for a certificate.
+func RevocationRecordName(cert ndn.Data) (enc.Name, bool) {
```
Comment (tianyuan129) [inline std/security/certificate.go]: What's the point of this function?

File: std/security/certificate.go [hunk 5]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
...
+	return revocationRecordName(cert)
+}
+
+// IsRevoked reports whether a revocation record Data packet is installed for the certificate.
+func IsRevoked(cert ndn.Data) bool {
```
Comment (tianyuan129) [inline std/security/certificate.go]: Same, all functions in this file should be stateless.

File: std/security/certificate.go [hunk 6]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
...
+	recordName, ok := revocationRecordName(cert)
+	if !ok {
+		return false
+	}
+
+	wire, _ := revocationStore.Get(recordName, false)
+	return len(wire) > 0
+}
+
+func encodeRevocationRecordContent(
```
Comment (tianyuan129) [inline std/security/certificate.go]: If I guess it right you are translating the function from the cxx repo into go. This library has its own way to define TLV and autogen encoding/decoding code. See rest of the library. Don't manually encode TLVs.

File: std/security/certificate.go [hunk 7]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
...
+	copy(out[pos:], value)
+	return out
+}
+
+func appendTlvNat(dst []byte, typ enc.TLNum, n uint64) []byte {
+	nat := enc.Nat(n)
+	return appendTlv(dst, typ, nat.Bytes())
+}
+
+func validateRevocationRecordData(data ndn.Data) error {
```
Comment (tianyuan129) [inline std/security/certificate.go]: I don't see the point of this function?

File: std/security/certificate.go [hunk 8]
```diff
@@ -115,6 +137,202 @@ func CertIsExpired(cert ndn.Data) bool {
...
+		return fmt.Errorf("revocation record is nil")
+	}
+	if !isRevocationRecordName(data.Name()) {
+		return fmt.Errorf("not a revocation record name: %s", data.Name())
+	}
+	_, err := parseRevocationRecordContent(data.Content())
+	return err
+}
+
+func parseRevocationRecordContent(content enc.Wire) (map[enc.TLNum][]byte, error) {
```
Comment (tianyuan129) [inline std/security/certificate.go]: Same, don't manually codec.

File: std/security/certificate.go [hunk 9]
```diff
@@ -125,15 +343,13 @@ func getPubKey(data ndn.Data) ([]byte, enc.Name, error) {
 
 	switch contentType {
 	case ndn.ContentTypeKey:
-		// Content is public key, return directly
```
Comment (tianyuan129) [inline std/security/certificate.go:128]: Same, why delete?

File: std/security/certificate_test.go [hunk 10]
```diff
@@ -13,7 +16,6 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-// (AI GENERATED DESCRIPTION): Tests that the `SignCert` function correctly returns an error when invoked with a nil signer and nil data.
```
Comment (tianyuan129) [inline std/security/certificate_test.go:16]: Same.

File: std/security/trust_config.go [hunk 11]
```diff
@@ -94,6 +93,28 @@ func (tc *TrustConfig) Suggest(name enc.Name) ndn.Signer {
 	return tc.schema.Suggest(name, tc.keychain)
 }
 
+// InstallRevocationRecord stores a revocation record Data packet in the keychain store.
+func (tc *TrustConfig) InstallRevocationRecord(wire enc.Wire) error {
```
Comment (tianyuan129) [inline std/security/trust_config.go]: How about `insertRevoke`?

File: std/security/trust_config.go [hunk 12]
```diff
@@ -94,6 +93,28 @@ func (tc *TrustConfig) Suggest(name enc.Name) ndn.Signer {
...
+	if !isRevocationRecordName(data.Name()) {
+		return fmt.Errorf("not a revocation record name: %s", data.Name())
+	}
+	if err := validateRevocationRecordData(data); err != nil {
+		return err
+	}
+
+	tc.mutex.Lock()
+	defer tc.mutex.Unlock()
+	return tc.keychain.Store().Put(data.Name(), wire.Join())
```
Comment (tianyuan129) [inline std/security/trust_config.go:114]: By design revocation records should not go into the keychain store...but realized we haven't had a better place. Please leave a comment there saying this is really a hack.

File: std/security/trust_config.go [hunk 13]
```diff
@@ -507,6 +543,44 @@ func (tc *TrustConfig) PromoteAnchor(cert ndn.Data, raw enc.Wire) {
...
+	if !ok {
+		return false
+	}
+
+	wire, _ := tc.keychain.Store().Get(recordName, false)
+	return len(wire) > 0
+}
+
+func (tc *TrustConfig) rejectIfRevokedCert(cert ndn.Data, callback func(bool, error)) bool {
+	if tc.certIsRevoked(cert) {
```
Comment (tianyuan129) [inline std/security/trust_config.go]: If this is the only call site, better inline the function.

File: std/security/trust_config.go [hunk 14]
```diff
@@ -507,6 +543,44 @@ func (tc *TrustConfig) PromoteAnchor(cert ndn.Data, raw enc.Wire) {
...
+	for _, root := range tc.roots {
+		if root.Equal(name) {
+			return true
+		}
+	}
+	return false
+}
+
+func (tc *TrustConfig) certIsRevoked(cert ndn.Data) bool {
+	if cert == nil || tc.isTrustedAnchorCert(cert) {
```
Comment (tianyuan129) [inline std/security/trust_config.go]: The function still works perfectly without `isTrustedAnchorCert`, given you will not install revocation record of trust anchors.

**Review by tianyuan129 — CHANGES_REQUESTED** [review]

File: std/security/certificate.go (hunk 1 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go:36]: Restored in f322b0a — sorry, a cleanup pass stripped those section comments. They are back on SignCert and getPubKey.

File: std/security/certificate.go (hunk 2 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Renamed to `RevokeCert` / `RevokeCertArgs` to match `SignCert`. Removed the old package-level `Revoke()`.

File: std/security/certificate.go (hunk 3 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Dropped the package-global `revocationStore`. `RevokeCert` only returns wire now; `TrustConfig.InsertRevoke` handles storage.

File: std/security/certificate.go (hunk 4 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Removed `RevocationRecordName`. Callers that need the name can read it from the built Data packet.

File: std/security/certificate.go (hunk 5 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Removed `IsRevoked` from certificate.go. Revocation checks now go through TrustConfig (keychain store lookup in `rejectIfRevokedCert`).

File: std/security/certificate.go (hunk 6 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Added `std/security/revocation_tlv/` with `gondn_tlv_gen` — same pattern as ndncert/tlv. Hand-rolled encode/parse helpers are gone.

File: std/security/certificate.go (hunk 7 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Removed `validateRevocationRecordData`. `InsertRevoke` only checks that the Data name matches `isRevocationRecordName`.

File: std/security/certificate.go (hunk 8 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Removed `parseRevocationRecordContent` along with the manual codec.

File: std/security/certificate.go (hunk 9 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go:128]: Restored the getPubKey switch comments in f322b0a.

File: std/security/trust_config.go (hunk 11 above)
Comment (Harsh23Kashyap) [reply std/security/trust_config.go]: Renamed to `InsertRevoke`.

File: std/security/trust_config.go (hunk 12 above)
Comment (Harsh23Kashyap) [reply std/security/trust_config.go:114]: Added a HACK comment above the `keychain.Store().Put` call — agreed this is not the right long-term home.

File: std/security/trust_config.go (hunk 13 above)
Comment (Harsh23Kashyap) [reply std/security/trust_config.go]: Inlined `certIsRevoked` into `rejectIfRevokedCert` (still called from the four Validate hooks).

File: std/security/trust_config.go (hunk 14 above)
Comment (Harsh23Kashyap) [reply std/security/trust_config.go]: Removed `isTrustedAnchorCert` and the anchor-skip test. We will not install anchor revocation records.

File: std/security/certificate_test.go (hunk 10 above)
Comment (Harsh23Kashyap) [reply std/security/certificate_test.go:16]: Sorry for the noise on the test comments — I read your "same" as restore everything that got deleted

Fixed in af39a3b: left short normal comments above the SignCert tests. SignCert/getPubKey section comments in certificate.go are unchanged.

**Comment by Harsh23Kashyap** [comment]
Pushed review fixes in f322b0a and follow-ups. Summary:

- `RevokeCert` / `RevokeCertArgs` — stateless, mirrors `SignCert`
- `revocation_tlv/` — gondn_tlv_gen for record content TLVs
- `TrustConfig.InsertRevoke` — stores records; Validate checks keychain store before using a cert
- HACK comment on keychain store Put as discussed

Tests pass locally (`go test -race ./std/security/...`). Ready for another look when you have time.

File: std/security/certificate_test.go (hunk 10 above)
Comment (Harsh23Kashyap) [reply std/security/certificate_test.go:16]: Restored those test comments in the latest push — should match revoke branch again.

File: std/security/certificate.go [hunk 15]
```diff
@@ -1,17 +1,41 @@
...
 
 	enc "github.com/named-data/ndnd/std/encoding"
 	"github.com/named-data/ndnd/std/ndn"
 	spec "github.com/named-data/ndnd/std/ndn/spec_2022"
+	revocationtlv "github.com/named-data/ndnd/std/security/revocation_tlv"
 	sig "github.com/named-data/ndnd/std/security/signer"
 	"github.com/named-data/ndnd/std/types/optional"
 )
 
+type revocationReason uint64
```
Comment (tianyuan129) [inline std/security/certificate.go]: use enum to enforce specific revocation reasons?

File: std/security/trust_config.go [hunk 16]
```diff
@@ -81,7 +81,6 @@ func NewTrustConfig(keyChain ndn.KeyChain, schema ndn.TrustSchema, roots []enc.N
 	}, nil
 }
 
-// (AI GENERATED DESCRIPTION): Returns the constant string `"trust-config"` for a `TrustConfig` value, enabling string formatting via the `fmt.Stringer` interface.
```
Comment (tianyuan129) [inline std/security/trust_config.go:84]: Why you delete this?

File: std/security/trust_config_test.go [hunk 17]
```diff
@@ -67,7 +67,6 @@ type SignCertOptions struct {
 	NotAfter  time.Time
 }
 
-// (AI GENERATED DESCRIPTION): Creates a signed certificate for the supplied Data packet using the provided signer, and returns the certificate wire, its content, and the signature‑covered portion.
```
Comment (tianyuan129) [inline std/security/trust_config_test.go:70]: Same. why delete?

**Review by tianyuan129 — CHANGES_REQUESTED** [review]
BTW #186 has been merged.

We need to resync this branch, and accommodate the signature timestamp. That is, if the signature timestamp presents and it is before the record notBefore, we should accept this data.

File: std/security/certificate.go (hunk 15 above)
Comment (Harsh23Kashyap) [reply std/security/certificate.go]: Added the typed enum.

**Comment by Harsh23Kashyap** [comment]
Implemented. Certs signed before record NotBefore are accepted.

File: std/security/trust_config.go (hunk 16 above)
Comment (Harsh23Kashyap) [reply std/security/trust_config.go:84]: Stripped down comment.

File: std/security/trust_config_test.go (hunk 17 above)
Comment (Harsh23Kashyap) [reply std/security/trust_config_test.go:70]: Removed as unnecessary.

**Review by tianyuan129 — APPROVED** [review]
LGTM. BTW we have lint issue, can you fix that?

@greetingsuniverse I gonna merge, hope this PR does not conflict with your SigTime fix?

**Comment by greetingsuniverse** [comment]
Doesn't look like there will be any conflicts, should be fine

**Comment by Harsh23Kashyap** [comment]
Yes, I have fixed that. The lint issue was a `=`-column alignment in the revocation-reason enum constants. Push is now in `3b7bbf9` ("goimports: align revocation reason enum") — gofmt is clean on the file and the rest of `std/security` still passes its race tests locally. Thanks for jumping on the suggestions-test fix; glad it does not clash with this branch.
